src/hooks/use-validation.js

```diff
@@ -0,0 +1,38 @@
+import { useLayoutEffect, useState } from 'react';
+import useMountedLayoutEffect from './use-mounted-layout-effect.js';
+
+// Custom hook for managing the validation message on a control.
+export default function useValidation(serverValidationMessage, inputRef) {
+  const [validationMessage, setValidationMessage] = useState(serverValidationMessage);
+
+  // On mount:
+  // Only set server validation message on the DOM node.
+  // The state should have been properly initialized.
+  useLayoutEffect(() => {
```

Comment:
Myabe combine?

Reply:
We can't do this actually.

If we combine them and run it on mount, calling `setValidationMessage` will bubble an invalid event and show the "required" fields as invalid on first time render. We want to avoid that as part of the design spec.